### PR TITLE
Fix quant type handling in stablehlo-legalize-quant-to-int pass

### PR DESCRIPTION
**Pull request by sdasgup3**
Originally brought up at https://github.com/openxla/stablehlo/pull/2383#discussion_r1630510134

The `stablehlo-legalize-quant-to-int` pass propagated from mho (#2383 ) uses std::variant to handle per-tensor and per-axis quantized type. However, the base class quant::QuantizedType provides a better polymorphic way to accomplish the same. 

## Direction to reviewers

Please review the change only in the following commit: https://github.com/openxla/stablehlo/pull/2385/commits/ee8b499fa691bbb7b6b3948bdc581dce05abaa39

Reason: This PR is based on https://github.com/openxla/stablehlo/pull/2383 which is not merged yet and hence has many files which are already reviewed as part of #2383.

### CODE REVIEW

**Comment by mlevesquedion** [comment]
Can you merge main and push? Will make reviewing a bit easier.

File: stablehlo/transforms/StablehloLegalizeQuantToInt.cpp [hunk 1]
```diff
@@ -0,0 +1,1338 @@
...
+    return quantType;
+  return failure();
+}
+
+bool isPerTensorType(Type type) {
+  return isa<quant::UniformQuantizedType>(getElementTypeOrSelf(type));
+}
+
+bool isPerChannelType(Type type) {
+  return isa<quant::UniformQuantizedPerAxisType>(getElementTypeOrSelf(type));
```
Comment (mlevesquedion) [inline stablehlo/transforms/StablehloLegalizeQuantToInt.cpp:64]: Should we align on one term, either "per channel" or "per axis"?

File: stablehlo/transforms/StablehloLegalizeQuantToInt.cpp [hunk 2]
```diff
@@ -0,0 +1,1338 @@
...
+
+// Extracts storage type of a UQ type. Return original type if it is no UQ type.
+Type getQuantStorageType(Type type) {
+  if (auto shaped = dyn_cast<ShapedType>(type)) {
+    return shaped.clone(getQuantStorageType(shaped.getElementType()));
+  }
+
+  auto quantizeType = getQuantType(type);
+  if (succeeded(quantizeType)) {
+    return getQuantStorageType(*quantizeType);
```
Comment (mlevesquedion) [inline stablehlo/transforms/StablehloLegalizeQuantToInt.cpp:157]: Nit: s/quantize/quant or s/quantize/quantized

**Review by mlevesquedion — APPROVED** [review]

File: stablehlo/transforms/StablehloLegalizeQuantToInt.cpp [hunk 3]
```diff
@@ -48,36 +48,28 @@ limitations under the License.
 namespace mlir::stablehlo {
 namespace {
 
-// TODO: b/311218165 - consider extract this to common utils and better ways to
-// handle polymorphism.
-using QuantType = std::variant<quant::UniformQuantizedType,
-                               quant::UniformQuantizedPerAxisType>;
+using QuantType = quant::QuantizedType;
```
Comment (mlevesquedion) [inline stablehlo/transforms/StablehloLegalizeQuantToInt.cpp:51]: IMO not worth aliasing, but up to you.

File: stablehlo/transforms/StablehloLegalizeQuantToInt.cpp (hunk 2 above)
Comment (sdasgup3) [reply stablehlo/transforms/StablehloLegalizeQuantToInt.cpp]: done

File: stablehlo/transforms/StablehloLegalizeQuantToInt.cpp (hunk 1 above)
Comment (sdasgup3) [reply stablehlo/transforms/StablehloLegalizeQuantToInt.cpp]: done.

File: stablehlo/transforms/StablehloLegalizeQuantToInt.cpp (hunk 3 above)
Comment (sdasgup3) [reply stablehlo/transforms/StablehloLegalizeQuantToInt.cpp:51]: let keep it as is mainly because of the extend it is used in the fine.

**Review by GleasonK — APPROVED** [review]

**Pull request merged**